patches/api/0416-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,758 @@
++        return PROVIDER.playerProfiles();
++    }
++
++    /**
++     * A block position argument.
++     *
++     * @return argument
++     */
++    @NotNull
++    public static ArgumentType<ArgumentResolver<BlockPosition>> blockPos() {
++        return PROVIDER.blockPos();
++    }
++
++    /**
++     * A signed message argument.
++     * This argument can be resolved to retreieve the underlying
++     * signed message.
++     *
++     * @return argument
++     */
++    @NotNull
++    public static ArgumentType<MessageArgumentResponse> signedMessage() {
++        return PROVIDER.signedMessage();
++    }
++
++
++    /**
++     * A component argument.
++     *
++     * @return argument
++     */
++    @NotNull
++    public static ArgumentType<Component> component() {
++        return PROVIDER.component();
++    }
++
++    @ApiStatus.Internal
++    interface VanillaArgumentProvider {
++
++        @NotNull
++        ArgumentType<ItemStack> itemStack();
++
++        @NotNull
++        ArgumentType<BlockState> blockState();
++
++        @NotNull
++        ArgumentType<MessageArgumentResponse> signedMessage();
++
++        @NotNull
++        ArgumentType<UUID> uuid();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Entity>> entity();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Player>> player();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Collection<Entity>>> entities();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Collection<Player>>> players();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<Collection<PlayerProfile>>> playerProfiles();
++
++        @NotNull
++        ArgumentType<ArgumentResolver<BlockPosition>> blockPos();
++
++        @NotNull
++        ArgumentType<Component> component();
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java b/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..a4cdba3967813337e4c3a0a9ba8b1cd9e2769be8
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/WrapperArgumentType.java
+@@ -0,0 +1,86 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import com.mojang.brigadier.StringReader;
++import com.mojang.brigadier.arguments.ArgumentType;
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Collection;
++
++/**
++ * An argument type that wraps around another argument.
++ * This argument is special in that the argument type that is wrapped will
++ * be sent to the client, not the wrapper itself.
++ *
++ * @param <T> type
++ * @param <B> base wrapping type
++ */
++@ApiStatus.Experimental
++public abstract class WrapperArgumentType<T, B> implements ArgumentType<T> {
```

Comment:
Is `Wrapper` the right word to describe this? I don't think we really need to expose the fact that these argument types "wrap" around a primitive type, just that they need to be converted from a primitive type. what about `CustomArgumentType` since its explicitly for the api to extend this for their own argument types. 

I just think the concept of "wrapping" and doesn't really need to be in the API.

If you insist on wrapping being here, I think it should be `WrappedArgumentType` instead